### PR TITLE
[Feature] Add zca whitening 

### PR DESCRIPTION
**Pull request by ceroytres**
Related to #434  #435 

Currently, I have a Pytorch/Sklearn hybrid ZCA whitening module called ZCAWhitening.

-  ZCAWhitening module has a forward method applies the whitening transform, and it has a fit and inverse_transform methods which function similar to sklearn's.
-  There is an option for including the fitting operation in the forward pass for two reasons:
    - Gradient checking will not pass because of the nature of how gradient check works (e.g.
      the forward function needs to be evaluated every time a single variable is checked).
    - Allowing for fitting in the forward pass can be potentially used in an implementation of  https://arxiv.org/abs/1804.08450 .
- There is an option to detach the computation of the transformations in order to avoid backprop them.

The functional zca_whitening_transforms just computes the whitening matrix and mean because it seems more useful to extract the transformation matrices and potential use them in torchvision.transforms.LinearTransform and the functional doesn't have persistence to apply to new data. 
 
I would appreciate feedback on the interface, because there was some gray areas in designing the interface. If all is good, I would like to tidy up some of the test and docs.

### CODE REVIEW

File: kornia/color/zca.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+        y: torch.Tensor = (x_flat).mm(self.T_inv) + self.mean
+
+        y = y.reshape(x.size())
+
+        return y
+
+
+def zca_whitening_transforms(inp: torch.Tensor, eps: float = 1e-7,
+                             biased: bool = False, compute_inv: bool = False) -> Tuple[torch.Tensor, ...]:
```
Comment (edgarriba) [inline kornia/color/zca.py]: I would rename to `kornia.zca_mean` or similar, in order to be consistent to pytorch functions like `torch.std_mean`
https://pytorch.org/docs/stable/torch.html#torch.std_mean

I propose this signtature:
```python
def zca_mean(input: torch.Tensor, unbiased: bool = True, return_inverse: bool = False, eps: float = 1e-7) -> Tuple[torch.Tensor, ...]:
```

File: kornia/color/zca.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        eps (float) : a small number used for numerial stablility.
+        biased (bool): Whether to use the biased estimate of the covariance matrix
+        compute_inv (bool): Whether to return the inverse transform matrix
+
+
+    returns:
+        Tuple[torch.Tensor, ...]:
+        A tuple containing the ZCA matrix and the mean vector, and if compute_inv = True, the
+        inverse whitening matrix is retured as well as the final return value.
+
```
Comment (edgarriba) [inline kornia/color/zca.py]: add expected shape and mini example

File: kornia/color/zca.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    N: int = inp.size()[0]
+    num_features: int = reduce(lambda a, b: a * b, inp.size()[1::])
+
+    mean: torch.Tensor = torch.mean(inp, dim=0, keepdim=True)
+    mean = mean.view((num_features,))
+
+    inp_center_flat: torch.Tensor = inp.view((N, num_features)) - mean
+
+    cov = inp_center_flat.t().mm(inp_center_flat)
```
Comment (edgarriba) [inline kornia/color/zca.py]: .t() won't worked in batched, same for .mm, use .transpose(-2, -1) and torch.matmul or .bmm to allow potenial broadcasting

File: kornia/color/zca.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    """
+
+    if not torch.is_tensor(inp):
+        raise TypeError("Input type is not a torch.Tensor. Got {}".format(
+            type(inp)))
+
+    N: int = inp.size()[0]
+    num_features: int = reduce(lambda a, b: a * b, inp.size()[1::])
+
+    mean: torch.Tensor = torch.mean(inp, dim=0, keepdim=True)
```
Comment (edgarriba) [inline kornia/color/zca.py]: @ceroytres what happens with batched tensors ? what's the expcted shape of for tensors (*, H, W) ?

File: kornia/color/zca.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    if biased:
+        cov = cov / float(N)
+    else:
+        cov = cov / float(N - 1)
+
+    U, S, _ = torch.svd(cov)
+
+    S = S.view(-1, 1)
+    S_inv_root: torch.Tensor = torch.rsqrt(S + eps)
+    T: torch.Tensor = (U).mm(S_inv_root * U.t())
```
Comment (edgarriba) [inline kornia/color/zca.py]: where did you find this implementation ? was checking this: https://martin-thoma.com/zca-whitening/

File: kornia/color/zca.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+    if compute_inv:
+        T_inv: torch.Tensor = (U).mm(torch.sqrt(S) * U.t())
+        return T, mean, T_inv
+    else:
```
Comment (edgarriba) [inline kornia/color/zca.py]: no need else statement:
```python
if return_inverse:
     return , mean, T_inv
return T, mean
```

File: kornia/color/zca.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
+from typing import Tuple
+from functools import reduce
+
+import torch
+import torch.nn as nn
+
+
+class ZCAWhitening(nn.Module):
```
Comment (edgarriba) [inline kornia/color/zca.py]: in my opinion, and for consistency to what we have been following, nn.Module should just be a wrapper around free functions that makes easy to be included nn.Sequential.

If possible I would like to have a free functional with the following signature:
```python
def zca_whitten(input: torch.Tensor, mean: Optional[torch.Tensor] = None, transform: Optional[torch.Tensor] = None) -> torch.Tensor:
    if mean is None and transform is None:
       transform, mean = zca_mean(input)
   # then apply
```
I'm not a huge fan of state functions, at least from the point of view of low level libs like this. I think the use can do it by itself. However, I'm totally open to iterate this proposal towards this direction.

**Comment by ceroytres** [comment]
Thanks for the feedback. I’m a bit sick right now and I have a deadline at the beginning of the month so I’ll probably won’t finish until sometime next month. Anyways take care!

**Comment by ducha-aiki** [comment]
Hi,

Any update on this?

**Comment by ceroytres** [comment]
I was hoping to start working on it again this upcoming week.

File: kornia/color/zca.py (hunk 1 above)
Comment (ceroytres) [reply kornia/color/zca.py:154]: Sure I can rename it.

File: kornia/color/zca.py (hunk 3 above)
Comment (ceroytres) [reply kornia/color/zca.py:191]: Maybe because I didn't mention it but the expected shape (N,D_1,D_2,...,D_n) (batch of N-D tensors). The tensors are flatten to a shape (N, D_1*D_2*...*D_n) (matrix) that's why I thought .t() or .bmm weren't needed. Hopefully this answers the next comment.

File: kornia/color/zca.py (hunk 5 above)
Comment (ceroytres) [reply kornia/color/zca.py:202]: Yeah. This implementation follows the exact Keras implementation. In my test, I compared to referenced implementation of https://github.com/kornia/kornia/issues/435#issue-558076239 .

File: kornia/color/zca.py (hunk 6 above)
Comment (ceroytres) [reply kornia/color/zca.py:207]: Makes sense. I'll remove this

File: kornia/color/zca.py (hunk 7 above)
Comment (ceroytres) [reply kornia/color/zca.py:8]: Yeah I agree. I think this was my number one point of contention when implementing this.

File: kornia/color/zca.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+        y: torch.Tensor = (x_flat).mm(self.transform_inv) + self.mean
+
+        y = y.reshape(x.size())
+
+        return y
+
+
+def zca_mean(inp: torch.Tensor, eps: float = 1e-7, biased: bool = False,
+             compute_inv: bool = False) -> Tuple[torch.Tensor, ...]:
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/color/zca.py:148]: Besides, `Tuple[torch.Tensor, ...]:` it's a bit confusing to understand what the function actually returns for someone that doesn't know. Then, we have to check what happen with JIT since the return type is variable (this is problem we might face in the whole library)

File: kornia/color/zca.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+    if compute_inv:
+        T_inv: torch.Tensor = (U).mm(torch.sqrt(S) * U.t())
+        return T, mean, T_inv
+
+    return T, mean
+
+
+def zca_whiten(input: torch.Tensor, mean: torch.Tensor,
+               transform: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/color/zca.py:210]: fro this signature, same thing. There a plan to include a group of `kornia.normalize_*` functions. This could be renamed to `kornia.normalize_zca` or `kornia.normalize_zca_whiten`.

File: kornia/color/zca.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+
+    returns:
+        (torch.Tensor) : Whiten Input data
+
+    Examples:
+        >>> x = torch.tensor([[0,1],[1,0],[-1,0]], dtype = torch.float32)
+        >>> transform, mean = zca_whiten(x)
+        >>> transform, mean, transform_inv = zca(x, mean, transform)
+
+    """
```
Comment (edgarriba) [inline kornia/color/zca.py:230]: add some check and shape type check

File: kornia/color/zca.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+    if mean is None and transform is None:
+        transform, mean = zca_mean(input)
```
Comment (edgarriba) [inline kornia/color/zca.py:233]: I would discourage this. To make the API more consistent a bug free I would assume the mean and transform has to be passed all the time. This is something that the user can easily do.

File: test/color/test_zca.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+        T, mu = kornia.color.zca_mean(data)
+
+        lt = LinearTransformation(T, mu)
+
+        out = torch.stack([lt(data[0]), lt(data[1]), lt(data[2]), lt(data[3])], axis=0)
+
+        assert_allclose(out, expected)
+
+    @pytest.mark.skip(reason="turn off all jit for a while")
```
[the remark this answers is not in the document]
Comment (ceroytres) [reply test/color/test_zca.py:159]: Yeah. I'll try this out.

File: kornia/color/zca.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ceroytres) [reply kornia/color/zca.py:148]: I'll check out how  ```Tuple[torch.Tensor, ...]```interacts with jit. I agree with the proposed changes for ```zca_mean``` since it would make switching between vanilla zca and the zca seen in decorrelated batchnorm easier. I'm assuming we'll apply similar argument changes to the module as well.

File: kornia/color/zca.py (hunk 10 above)
Comment (ceroytres) [reply kornia/color/zca.py:230]: Are saying add it to the example in the comment or the body of the function or both?

File: kornia/color/zca.py (hunk 11 above)
Comment (ceroytres) [reply kornia/color/zca.py:233]: I was following this suggestion https://github.com/kornia/kornia/pull/458#discussion_r392694327 but I agree. Should we change the name from zca_whiten to linear_transformation ([similar to Torchvision](https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision.transforms.LinearTransformation)) since without the zca_mean in the function, functionality of the function is similar to LinearTransformation from Torchvision?

File: kornia/color/zca.py (hunk 10 above)
Comment (edgarriba) [reply kornia/color/zca.py:230]: Just in the function, minimum raise error with some explicative message in case the user inputs something that's not supposed to

File: kornia/color/zca.py (hunk 11 above)
Comment (edgarriba) [reply kornia/color/zca.py:233]: I think `normalize_zca` should exist. Probably by just removing mean and transform from signature and always calling `mean_zca` + `linear_transform`. In this way we'll introduce this three new functionalities. As for linear_transform, do they really support batch in torchvision? Just to have a clear idea of what would be the difference between functionalities. I see they target this for specific use case, in our case we should think to make it more generic to be used for others use cases. Besides, this last one should end up in kornia.geometry.linalg. Last, check this linear operations group functions to see if we can get any optimization for this https://pytorch.org/docs/stable/torch.html?highlight=bmm#torch.addbmm

File: test/color/test_zca.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ceroytres) [reply test/color/test_zca.py:159]: Hey @edgarriba , I was taking a look at the jit stuff. It seems ```torch.is_tensor``` is problematic for ```torch.jit.script```. I added a [Colab Notebook](https://colab.research.google.com/drive/1stYuhcoRTradbRm9HqJlcBUZqJXYnier) showing the problem. It's odd since  ```torch.is_tensor```  is essentially ```isinstance(x, torch.Tensor)``` according to [here](https://pytorch.org/docs/stable/_modules/torch.html#is_tensor)

File: kornia/color/zca.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ceroytres) [reply kornia/color/zca.py:148]: After looking into it a bit more, I have some lingering question and clarifications. Assuming the dim arg denotes the sample/batch axis, dim = 0, keepdim=True,and following similar output to torch.std_mean, for an input tensor of shape (2,3,4,5), the mean has a shape (1,3,4,5). Should the zca transformation be returned as a (60,60) matrix and keepdim only affects the shape of the mean?

File: kornia/color/zca.py (hunk 11 above)
Comment (ceroytres) [reply kornia/color/zca.py:233]: I agree ```normalize_zca``` should exist and maybe it should be bundle ```mean_zca+linear transform```. Torchvision's [linear transform](https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision.transforms.LinearTransformation) doesn't handle batch since the use case focuses on transforms in dataloaders. Agreed a linear transform function should end up in in kornia.geometry.linalg . I'll take a look at ```torch.addbmm``` seems extremely useful.

File: test/color/test_zca.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply test/color/test_zca.py:159]: you are right :)
actually JIT author suggested to use isinstance since would be more pythonic, otherwise we can send a PR directly to cover this ops in JIT. Would go for pythonic way.

Besides, I have a pending work to create a minimal API in kornia.testing to have different `check_*` functions. `check_is_tensor` would be the first one in the list :)

[the remark this answers is not in the document]
Comment (edgarriba) [reply test/color/test_zca.py:159]: `check_type` could be another one that receives multiple possible input type

File: kornia/color/zca.py (hunk 11 above)
Comment (edgarriba) [reply kornia/color/zca.py:233]: cool. In that case makes sense to support batched version.

File: kornia/color/zca.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/color/zca.py:148]: probably yes. In this case I guess the output should be specified by (B, C*H*W). /cc @ducha-aiki any suggestions in this ?

**Comment by ceroytres** [comment]
@edgarriba I still needed to correct some of zca the documentation. I didn't include jit test float16 and float64 as suggested in https://github.com/kornia/kornia/pull/458#discussion_r416407397 . Also did you want me to include documentation for my linear_transform function as well https://github.com/ceroytres/kornia/blob/e77627d703549a42024e2d187eb7c9f1d4f25ad7/kornia/color/zca.py#L290. I believe I still have a minor bug in one of the test https://github.com/ceroytres/kornia/blob/e77627d703549a42024e2d187eb7c9f1d4f25ad7/test/color/test_zca.py#L55. Should I create a new pull request for these changes? Sorry if there was any confusion.

**Comment by edgarriba** [comment]
Sure, go ahead and open a new PE for those changes.

**Comment by ceroytres** [comment]
@edgarriba I created a PR #566. I think this might be sufficient for right now.

**Comment by Kshitij09** [comment]
@ceroytres how close the `inverse_transform` will be from original Tensor?
```python
x,*_ = dls.one_batch()
zca_block = ZCAWhitening(compute_inv=True).fit(x)
x_whiten = zca_block(x)
x_inv = zca_block.inverse_transform(x_whiten)
torch.allclose(x_inv,x,atol=1e-4)
```
Output: ```False```

Am I doing anything wrong?

**Comment by Kshitij09** [comment]
Also, does this output looks good? i applied `ZCAWhitening` on cifar-10 dataset
**Before Transform** 
image

**After Transform** (Applied min-max scaling on the output Tensor)
image

**Comment by ceroytres** [comment]
Yeah the output seems reasonable for the pictures according to this: https://www.kdnuggets.com/2018/10/preprocessing-deep-learning-covariance-matrix-image-whitening.html/3
They seem to use eps=.1

What’s your batch size? Typically if the batch size is smaller than the dimension of the data things tend to get numerically unstable since the covariance matrix isn’t full rank. This might be the reason why inverse isn’t exact. I’ll double check that this weekend.  https://stats.stackexchange.com/questions/60622/why-is-a-sample-covariance-matrix-singular-when-sample-size-is-less-than-number/180360#180360

**Comment by Kshitij09** [comment]
My batch size is `64`. Will try with higher batch dimension

**Comment by ceroytres** [comment]
Also I forgot to mention that you could use a larger eps since it acts like a regularizer.
